File: 2024/24/24a.py
INPUT_FILE = "main_input.txt"

def main():
    with open(INPUT_FILE) as f:
        initialVals, config = f.read().split("\n\n")
    iV = {}

    for line in initialVals.split("\n"):
        key, val = line.split(": ")
        iV[key] = int(val)

    zWires = set()
    pendingLines = config.split("\n")

    while len(pendingLines) > 0:
        localLines = []
        for line in pendingLines:
            inp, out = line.split(" -> ")
            a, op, b = inp.split(" ")
            if iV.get(a) == None or iV.get(b) == None:
                localLines.append(line)
                continue
            if op == "AND":
                iV[out] = iV[a] & iV[b]
            elif op == "OR":
                iV[out] = iV[a] | iV[b]
            elif op == "XOR":
                iV[out] = iV[a] ^ iV[b]

            if out.startswith("z"):
                zWires.add(out)
        pendingLines = localLines

    res = 0
    i = 0
    sortedZWires = sorted(zWires)
    for wire in sortedZWires:
        print(wire, iV[wire])
        res += iV[wire] << i
        i += 1
    return res

print(main())

File: 2024/24/24a.ts
const INPUT_FILE = Deno.args[0] === "prod" ? "./main_input.txt" : "./input.txt";

const main = () => {
  const [initialVals, config] = Deno.readTextFileSync(INPUT_FILE).split("\n\n");
  const iV: Record<string, number> = {};

  initialVals.split("\n").forEach((line) => {
    const [key, val] = line.split(": ");
    iV[key] = Number(val);
  });

  const uzWires = new Set<string>();
  let configs = config.split("\n")

  while (configs.length > 0) {
    const localConfig: string[] = [];
    configs.forEach((line) => {
      const [inp, out] = line.split(" -> ");
      const [a, op, b] = inp.split(" ");
      if (iV[a] === undefined || iV[b] === undefined) {
        localConfig.push(line);
        return;
      }

      if (op === "AND") {
        iV[out] = iV[a] & iV[b];
      } else if (op === "OR") {
        iV[out] = iV[a] | iV[b];
      } else if (op === "XOR") {
        iV[out] = iV[a] ^ iV[b];
      }

      if (out.startsWith("z")) {
        uzWires.add(out);
      }
    });
    configs = localConfig;
  }

  const zWires = [...uzWires].sort();
  let res = 0;
  let i = 0;
  for (const wire of zWires) {
    console.log(wire, iV[wire]);
    res += iV[wire] << i;
    i++;
  }
  return res;
}

console.time("Time");
console.log(main());
console.timeEnd("Time");
